src/VecSim/spaces/computer/computer.h

```diff
@@ -0,0 +1,309 @@
+        : IndexComputerAbstract<DistType>(allocator), alignment(alignment),
+          distance_calculator(distance_calculator) {}
+
+    ~IndexComputerBasic() override {
+        if (distance_calculator) {
+            delete distance_calculator;
+        }
+    }
+
+    ProcessedBlobs preprocess(const void *original_blob,
+                              size_t processed_bytes_count) const override;
+
+    MemoryUtils::unique_blob preprocessForStorage(const void *original_blob,
+                                                  size_t processed_bytes_count) const override;
+
+    unsigned char getAlignment() const override { return alignment; }
+
+    MemoryUtils::unique_blob preprocessQuery(const void *original_blob,
+                                             size_t processed_bytes_count) const override;
+
+    void preprocessQueryInPlace(void *blob, size_t processed_bytes_count) const override;
+
+    DistType calcDistance(const void *v1, const void *v2, size_t dim) const override {
+        assert(this->distance_calculator);
+        return this->distance_calculator->calcDistance(v1, v2, dim);
+    }
+
+protected:
+    const unsigned char alignment;
+    DistanceCalculatorAbstract<DistType, DistFuncType> *distance_calculator;
+
+    // Allocate and copy the blob only if the original blob is not aligned.
+    MemoryUtils::unique_blob maybeCopyToAlignedMem(const void *original_blob,
+                                                   size_t blob_bytes_count) const;
+
+    MemoryUtils::unique_blob wrapAllocated(void *blob) const {
+        return MemoryUtils::unique_blob(
+            blob, [this](void *ptr) { this->allocator->free_allocation(ptr); });
+    }
+
+    static MemoryUtils::unique_blob wrapWithDummyDeleter(void *ptr) {
+        return MemoryUtils::unique_blob(ptr, [](void *) {});
+    }
+};
+
+template <typename DistType, typename DistFuncType, size_t n_preprocessors>
+class IndexComputerExtended : public IndexComputerBasic<DistType, DistFuncType> {
+protected:
+    std::array<PreprocessorAbstract *, n_preprocessors> preprocessors;
+
+public:
+    IndexComputerExtended(
+        std::shared_ptr<VecSimAllocator> allocator, unsigned char alignment,
+        DistanceCalculatorAbstract<DistType, DistFuncType> *distance_calculator = nullptr)
+        : IndexComputerBasic<DistType, DistFuncType>(allocator, alignment, distance_calculator) {
+        assert(n_preprocessors);
+        std::fill_n(preprocessors.begin(), n_preprocessors, nullptr);
+    }
+
+    ~IndexComputerExtended() override {
+        for (auto pp : preprocessors) {
+            if (!pp)
+                break;
+
+            delete pp;
+        }
+    }
+
+    /** @returns On success, next uninitialized index, or 0 in case capacity is reached (after
+     * inserting the preprocessor). -1 if capacity is full and we failed to add the preprocessor.
+     */
+    int addPreprocessor(PreprocessorAbstract *preprocessor);
```

Comment:
We always know at compilation time the size of the array so we can save the memory overhead of a dynamic array.
To avoid checks we can save the current number of initiazlized preprocessor, but assuming we are not expecting to have a large amount of preprocessors, i think it won't make any diffrence.